packages/graph/hash_graph/lib/graph/src/store/postgres/resolve.rs

```diff
@@ -0,0 +1,161 @@
+use async_trait::async_trait;
+use error_stack::{IntoReport, Result, ResultExt};
+use tokio_postgres::{GenericClient, Row, RowStream};
+use type_system::{uri::VersionedUri, DataType, PropertyType};
+
+use crate::store::{postgres::parameter_list, AsClient, PostgresStore, QueryError};
+
+#[async_trait]
+pub trait PostgresContext {
+    async fn read_all_data_types(&self) -> Result<RowStream, QueryError>;
+
+    async fn read_versioned_data_type(
+        &self,
+        uri: &VersionedUri,
+    ) -> Result<Record<DataType>, QueryError>;
+
+    async fn read_all_property_types(&self) -> Result<RowStream, QueryError>;
+
+    async fn read_versioned_property_type(
+        &self,
+        uri: &VersionedUri,
+    ) -> Result<Record<PropertyType>, QueryError>;
+}
+
+#[async_trait]
+impl<T> PostgresContext for &T
+where
+    T: PostgresContext + Sync,
+{
+    async fn read_all_data_types(&self) -> Result<RowStream, QueryError> {
+        PostgresContext::read_all_data_types(*self).await
+    }
+
+    async fn read_versioned_data_type(
+        &self,
+        uri: &VersionedUri,
+    ) -> Result<Record<DataType>, QueryError> {
+        PostgresContext::read_versioned_data_type(*self, uri).await
+    }
+
+    async fn read_all_property_types(&self) -> Result<RowStream, QueryError> {
+        PostgresContext::read_all_property_types(*self).await
+    }
+
+    async fn read_versioned_property_type(
+        &self,
+        uri: &VersionedUri,
+    ) -> Result<Record<PropertyType>, QueryError> {
+        PostgresContext::read_versioned_property_type(*self, uri).await
+    }
+}
+
+#[derive(Debug)]
+pub struct Record<T> {
```

Comment:
These records are used for the expression filtering process, could we add doc strings to capture that? :)

Reply:
I added a short description. I will expand this when doing the documentation for the AST ([Asana Task](https://app.asana.com/0/0/1202884883200976/f) (_internal_))